HW-2 Model 적용

sangjinKIM93 opened the pull request:
HW-2 Model 적용하기

1. Repository 구조 적용 : Repository, source(remote, local), model
2. Repository에서 local cache 적용

*RemoteDataSource :  Retrofit2 사용
*LocalDataSource : Room 사용
*RxJava 추가(RoomDB 접근시 사용)

Commit:
RemoteDataSource & Impl 생성

Commit:
NaverMoviesRepository & Impl 생성 (only remote version)

아직은 remoteDataSource만 받아올 수 있습니다.

Commit:
data Model 리팩토링 & room 라이브러리 추가

Commit:
RoomDB 클래스, MovieDao 추가

Commit:
LocalDataSource & Impl 생성

Commit:
NaverMovieRepository에 localDataSource 추가

이와 더불어 필요한 객체인 roomDB도 생성자에 추가

Commit:
remote 통신 성공시 local에 cache 저장하는 로직 작성

Commit:
remote 성공시 local에 cache 저장하는 로직 작성(수정 1)

repository의 saveMoviesCache메서드 내부 구현을 누락하여 추가 작성하였습니다.

Commit:
RxJava 라이브러리 추가 및 RoomDB 접근시 RxJava 활용

LocalDataSource에서 RoomDB 접근을 위해 RxJava를 활용했습니다.

Commit:
Activity 수정

기존 레트로핏에서 직접 결과를 받아 처리하던 것에서
Repository에서 결과를 받아 처리하는 것으로 변경하였습니다.

Commit:
Movie.class 패키지 경로 변경

이전 커밋에서 data model하면서 리팩토링 해줬었는데 패키지 경로 변경이 진행되지 않았던 부분이 있었습니다.

RyuC0605 commented on 2005/sangjinKIM93/newproject/app/src/main/java/com/sangjin/newproject/data/source/local/MovieDao.kt:
```diff
@@ -0,0 +1,25 @@
+    @Insert(onConflict = OnConflictStrategy.REPLACE)
+    fun insertAll(movies: List<Movie>)
+
+    @Query("SELECT * FROM movie_table")
+    fun getAllMovies() : List<Movie>
```
SELECT * 보다 사용하는 컬럼만 셀렉트 하는것이 여러모로 좋습니다.

sangjinKIM93 replied:
e718052
특정 칼럼만 선택하여 select 해보았는데 다음과 같은 에러가 발생하였습니다.

The columns returned by the query does not have the fields [movieId,subtitle,pubDate,director,actor,userRating] in com.sangjin.newproject.data.model.Movie even though they are annotated as non-null or primitive. Columns returned by the query

관련 내용을 찾아보니 room에서는 movie객체를 return하기 위해서는 movie data class에 선언된 모든 column들이 select문에 포함되어야 하는 것 같습니다. 
그래서 movie data class에서 불필요한 column을 삭제하는 방식으로 수정하였는데 혹시 더 좋은 생각 있으시다면 추가적으로 반영하겠습니다!

RyuC0605 replied:
사용하지 않는 컬럼을 만들어서 얻는 이득이 있을까요?

sangjinKIM93 replied:
그렇죠 필요한 칼럼만 사용하는게 좋은 것 같습니다:)

RyuC0605 commented on 2005/sangjinKIM93/newproject/app/src/main/java/com/sangjin/newproject/data/source/local/MovieDao.kt:
```diff
@@ -16,8 +17,8 @@ interface MovieDao {
-    @Query("SELECT * FROM movie_table WHERE title = :query")
-    fun selectMovies(query: String): List<Movie>
+    @Query("SELECT * FROM movie_table WHERE title LIKE :query")
+    fun selectMovies(query: String): Single<List<Movie>>
```
역시 SELECT * 사용은 지양하는게 좋습니다.

sangjinKIM93 replied:
e718052
위와 동일합니다.

RyuC0605 reviewed (changes requested):
안녕하세요. 류승윤 입니다.
즐거운 한주 보내셨는지요? 
그럼 리뷰 시작 하겠습니다.

Context 는 전역변수로 두지 않는 것이 좋은데 MovieListAdapter 에 그런 코드가 있네요.
그리고 같은 클래스의 onBindViewHolder 함수의 holder.containerView.tag = getItemId(position)
코드는 왜 존재하는지 알려주시겠어요?

동작이슈 
검색어에 "너나다" 를 입력하면 검색 결과가 있는데, 검색결과가 없다고 토스트메시지가 나오네요.

Commit:
불필요한 tag, context 전역변수 삭제

+clickListener 코드를 좀 더 깔끔하게 수정했습니다.

Commit:
검색결과가 있는데 "검색결과가 없습니다" 토스트창 뜨는 에러 해결

(원인)
local 캐시를 먼저하고 remote 통신을 해서 최신화 하는 구조인데
local 캐시 과정에서 관련 내용이 없지만 remote에서는 관련 검색 내용이 있는 경우 발생한 에러.
(해결방법)
local에서 검색한 내용이 NullOrEmpty일 경우 콜백을 반환하지 않도록 함.(in LocalDataSourceImpl)

Commit:
Room select 할 때 불필요한 column 문제 해결

data class에서 불필요한 column을 삭제함

sangjinKIM93 commented:
안녕하세요 승윤님
이번에도 좋은 리뷰 감사합니다:)


925f25a
1. tag는 지난번 수정때 미처 삭제를 못하였네요. 삭제하였습니다.

2. context는 glide때문에 전역으로 설정하였었는데 viewholder의 context를 직접 받는 방식으로 수정하였습니다.


cb64d9a
3. Toast 메시지 에러는 local cache에서 자료가 없을때  onSuccess콜백으로 넘어와서 생기는 에러였습니다.  local에서 select한 결과가 NullOrEmpty일 경우 onSuccess 콜백이 넘어가지 않도록 예외처리하였습니다.

Commit:
interface로 roomDB 넘겨주던 방식 -> localDataSource 매개변수로 받는 것으로 변경

Commit:
불필요한 interface 삭제 (saveMoviewCache)

private 메서드로 변경

Commit:
Local Cache 사용법 변경. "초기 화면 로드시 최근 검색한 결과를 보여준다."

기존에는 검색시 그 내용을 local에 저장하고 이후 같은 키워드로 검색을 했을떄 local에 있는 데이터를 먼저 불러오는 형식이었는데 불필요한 작업이라고 생각되어 '초기 화면 로드시 최근 검색 결과를 보여주는 기능'으로써 local 캐시를 활용함

Commit:
RemoteDataSource에서 subscribe()하던 것에서 Activity에서 subscribe()하는 것으로 변경

onDestroy에서 disposable을 해체

sangjinKIM93 commented:
추가적인 변경사항입니다.

1. 기존 캐시 방식(검색시 local에 자료가 있을 경우 불러오는 구조)이 부적절한 것 같아. 앱 초기 화면에 최근 검색한 자료를 불러오는 데 local cache를 활용하는 것으로 변경하였습니다.

2. disposable 처리를 위해 기존 DataSource에서 subscribe()하던 것을 activity에서 subscribe 하도록 변경하였습니다.

sjjeong commented on 2005/sangjinKIM93/newproject/app/src/main/java/com/sangjin/newproject/data/repository/NaverMoviesRepositoryImpl.kt:
```diff
@@ -0,0 +1,28 @@
+import io.reactivex.Single
+
+class NaverMoviesRepositoryImpl(
+    private val remoteDataSourceImpl: RemoteDataSourceImpl,
+    private val localDataSourceImpl: LocalDataSourceImpl
```
타입은 interface 형태로 변경해주세요

sangjinKIM93 replied:
c71a0e7

수정했습니다.

이렇게 타입을 바꿈으로써 얻는 장점이 뭔지 고민해봤는데 잘 모르겠네요 ㅠ
interface로 받는 것과 Impl로 받는 것에 어떤 차이점이 있는지 알려주실 수 있나요??

sjjeong replied:
보통 interface를 사용하는 이유는 다형성을 위해서 입니다.
하지만 여기서 사용하는 것은 추후 모듈화 때 사용하기 위해서 입니다.

sangjinKIM93 replied:
presenter 계층이 추가되었을때를 위해서 사용한다는 말씀이신가요??

sjjeong commented on 2005/sangjinKIM93/newproject/app/src/main/java/com/sangjin/newproject/data/source/local/LocalDataSourceImpl.kt:
```diff
@@ -0,0 +1,26 @@
+import io.reactivex.android.schedulers.AndroidSchedulers
+import io.reactivex.schedulers.Schedulers
+
+class LocalDataSourceImpl(room : RoomDB) : LocalDataSource {
+
+    private val roomDB = room
```
생성자 쪽에 선언해주면 됩니다.

sangjinKIM93 replied:
54a072e
감사합니다. 수정했습니다.

sjjeong commented on 2005/sangjinKIM93/newproject/app/src/main/java/com/sangjin/newproject/data/source/local/RoomDB.kt:
```diff
@@ -0,0 +1,42 @@
+import com.sangjin.newproject.data.model.Movie
+
+@Database(entities = arrayOf(Movie::class), version = 2)
+abstract class RoomDB : RoomDatabase(){
```
클래스명을 RoomDb로 변경해주세요

sangjinKIM93 replied:
fc2a7f5

수정했습니다:)

sjjeong reviewed (changes requested):
model layer 구현은 잘 하셨네요

Commit:
Local에서 roomDb를 생성자쪽에서 생성해주는 것으로 변경

Commit:
interface 타입으로 변경

Commit:
RoomDB -> RoomDb로 수정

sjjeong reviewed (approved):
LGTM ✊ ✌️ 👋
